### PR TITLE
Windows10 UWP

### PR DESCRIPTION
**Pull request by perneky**
Hi Henrik,

This is the port of PPSSPP for Windows 10 (PC, phone, and XboxOne). Maybe things are not always in the correct place, but it is working. You can compile it under Windows 10, by using the Win32 UWPDebug/UWPRelease configs.

My main concern is two areas, the memmap.cpp memory model, as MapViewOfFileEx misses the base argument on Windows 10 (MapViewOfFileFromApp), so I try an other approach in memmap.

The other is JIT, while there is VirtualProtectFromApp, PAGE_EXECUTE_READWRITE is not allowed, so I keep the executable memory EXECUTABLE, and while JITting, I change it to READWRITE, then EXECUTABLE again.

Can you please tell me your opinion?

Laci

### CODE REVIEW

**Comment by hrydgard** [comment]
Hi @perneky ,

Thank you for working on this! Looks like a very big change, so I won't have time to review it properly tonight. Will look at it in the upcoming days.

**Comment by unknownbrackets** [comment]
This pull is pretty big.  A few things:

1. UWP/Content/ - this seems to be a copy of assets/.  We shouldn't track two copies of the same files.  Instead, we should use a build script (prebuild if you like, even.)  Just like Android, it can copy the files needed into the right place.

2. Some of the indentation is off: https://github.com/hrydgard/ppsspp/pull/8315/files#diff-2b8fbb5b62261b971fe141ef565cdbafL126 - please try to keep consistent with tabs.

3. This is a large set of changes, so it's likely there will a bug hiding somewhere.  If we merge it all at once, and something goes wrong, we'll probably be virtually unable to debug it - and left with (a) reverting the whole pull request, (b) leaving whatever bug in, or (c) trying to fix it and likely breaking this Windows 10 support.

It'd be better to break changes up.  For example, you could create a new branch from master, and collect things like [ifdefing out DX9] (https://github.com/hrydgard/ppsspp/pull/8315/files#diff-20fe7d16498e76424a83bd8cf44365feR31), [alternate Windows APIs for folders and stuff] (https://github.com/hrydgard/ppsspp/pull/8315/files#diff-130401d995253e54392d8b71dbad2eceR584), etc. into a few commits in one pull request.

Then we could look at this pull request, which won't be enough yet to build or run on Windows 10 yet, and merge it.  If things go wrong, we'll be in a much better place to deal with the problem.

By doing it in chunks, we can move forward a lot more steadily and efficiently.  I know it's a pain to break out commits, though, after the fact.

4. There are some places that have clearly been added to make things work, but aren't correct, [like this one] (https://github.com/hrydgard/ppsspp/pull/8315/files#diff-a2f55e0ac27d84017e8d6bfc15d62d66R300) and [this one] (https://github.com/hrydgard/ppsspp/pull/8315/files#diff-a92938ecb8eb753648b3d9658dc0aed0R27).  There are [other changes that seem questionable] (https://github.com/hrydgard/ppsspp/pull/8315/files#diff-29313f9f6ed28d0e24e6dac4e0db8933R1125) or [probably temporary?] (https://github.com/hrydgard/ppsspp/pull/8315/files#diff-81225357fb059f16c55ad743d356f41bR339)  But a huge list of comments on this pull will be hard to manage, so that's another reason to separate it into chunks.

5. ffmpeg_UWP belongs inside the ffmpeg/ repo, just like all other platforms.

Anyway, glad to hear you got it working - that's certainly a lot of hard work.  I can't tell yet, but is jit even working?  Does Windows 10 want everything to run in Thumb mode?

-[Unknown]

**Comment by perneky** [comment]
Hi Guys,

I know that this version still needs work before it can be merged to a production branch. There are lots of temporary and hack solutions in the code, and also some changes I have tried to address some issues, in other words, I know that I have to clean this up.

The main reason why I sent the pull request as it is now to get some feedback especially from the memmap, the memarena and the memoryutil changes.

I wasn't able to use the memory handling from the windows version, as the MapViewOfFileFromApp function is missing the base parameter, so I have tried an other approach, allocating for the views.

JIT is working. Instead of VirtualProtect, we have VirtualProtectFromApp, which is basically the same but not allows for memory to be executable and writable at the same time. So for that I keep the memory executable, and before JITting, I make the next area writable, and after JITing I make it executable again. Maybe this could be done better, I am hoping for some feedback about this solution as well, you can search for MemoryAccess to see how and where I do this.

In general, the x86 version is sort of the same as the desktop version, but a bit slower. The ARM version is a bit slower as well (Lumia 930 vs Nexus 6), but it can be because the nexus has better hardware. There are two main problems left. One is that some games like LBP, are crashing on ARM, but not on x86, and also not on Android. The crash is always illegal memory access, so I am suspecting that my solution for the memory handling has a flaw. The other problem is that on ARM some 2D drawings are missing. Like the faces of LocoRoco, or the flame of the dragon in PataPon. Those are mostly textured with alpha. It seems that the matrices for those renderings are invalid, but I haven't gone deep into that.

I am hoping that by getting a review on the memory and maybe some other areas, those things can be fixed.

Thank you.

Laci

**Comment by unknownbrackets** [comment]
Well, it's hard to review this code with all the other stuff, as mentioned.

To clarify on memory mapping, one important thing is how memory mirrors work.  Basically, there are two memory addresses which both access the same memory (actually more, in some cases, but let's keep it simple):

0x08800000
0x48800000

So, if I write "42" to the second address, and then read from the first address, I should get 42.  The 0x40000000 tells it to read without using the memory cache, but we don't support that, so they are the same operation for us.

To make this work, and to make this fast, we construct a `base` pointer, and then we make it so that `base + 0x08800000` and `base + 0x48800000` both actually point to the exact same memory.

I don't really know for sure whether the Xbox 360 code worked or not.  But in general, are you constructing a view of memory where the above is correct?

-[Unknown]

**Comment by perneky** [comment]
Thank you for clarification. Mirroring of the area is not possible in windows 10, as the MapViewOfFileEx function is missing. There is MapViewOfFileFromApp, which is basically the same, but missing the last base parameter, which would make this possible.

What I do is to create the base memory:

base = (u8*)VirtualAllocFromApp(0, 0x10000000, MEM_RESERVE|MEM_COMMIT, PAGE_READWRITE);

Then for the views, I create blocks of memory for the views array:

if ( !CanIgnoreView( view ) ) {
    if ( view.out_ptr_low )
        *( view.out_ptr_low ) = (u8*)( base + view.virtual_address );
    ptr = VirtualAllocFromApp( base + ( view.virtual_address & MEMVIEW32_MASK ), view.size, MEM_COMMIT, PAGE_READWRITE );
    }
*( view.out_ptr ) = (u8*)base + ( view.virtual_address & MEMVIEW32_MASK );

It is basically the same as for xbox. Strange thing that it is working for x86 really well, I can run lots of games. But on ARM, it is crashing with invalid memory access (load/store) error for many games, even games that is working on Android.

Does my setup looks ok, or could show me a version which should work without MapViewOfFileEx?

Thank you.

Laci

**Comment by hrydgard** [comment]
Using MEMVIEW32_MASK eliminates the need for the mirroring, at a small performance penalty. We don't do the mirroring in 32-bit builds anyway currently, as there's usually not enough VM space, so that's fine.

Not sure why it would be crashing on ARM though, but it may have other causes, for example the ARM calling convention may be subtly different on Windows somehow...

**Comment by unknownbrackets** [comment]
Well, MEMVIEW32_MASK won't work for VRAM.

-[Unknown]

**Comment by hrydgard** [comment]
Yeah, that's true ..

File: Common/CPUDetect.cpp [hunk 1]
```diff
@@ -65,6 +65,7 @@ static unsigned long long _xgetbv(unsigned int index)
 #endif
 
 #if !defined(MIPS)
+#if !( defined(_WIN32) && defined(_M_ARM) )
```
Comment (unknownbrackets) [inline Common/CPUDetect.cpp]: ARM simply should not compile this file.  It should instead only compile ArmCPUDetect.cpp.

-[Unknown]

**Comment by hrydgard** [comment]
So, it just became possible for everyone to turn their Xbox Ones into devkits, which creates a nice new usecase for this.

This branch currently uses ANGLE to convert OpenGL to Direct3D11, but as Xbox One supports DX12, it might make sense to create such a backend instead, once Vulkan is done (DX12 is similar enough to Vulkan that a port that way should be pretty quick). But of course we can start with ANGLE.

Just not sure how to approach merging this, as has been said, it really needs to be split up into commits.

**Comment by GAMELASTER** [comment]
btw, is possible to build into a W10M?

**Comment by Megasware128** [comment]
It is. I have tried that. But it did crash for me with GTA: LCS while it does run on W10 itself. Not sure about other games.

**Comment by annahana** [comment]
will it work with continium on windows 10 mobile?

**Comment by Megasware128** [comment]
It should work since it's just an UWP app. But if u want to I can test it since I got a display dock over here.

**Comment by MobWiMetro** [comment]
@perneky In the past, Windows on ARM allows code to switch from Thumb-2 to ARM, but will crash on Interrupt. Isn't the memory access failure caused by this?

**Comment by perneky** [comment]
Hi all,

Originally I have started to work on this port to make it possible to use the emulator on my phone. I have been working a lot on this one, this is clearly not a final stage, but a work in progress version. My original intent with this request was to get feedback. It works on x86 and ARM based Windows 10 devices, but not as much as on Android or Windows Desktop. Clearly the main issue is the corrupt memory access. That was the main thing I needed help with. In UWP, you can not use memory mapped files the way the desktop version is using them. I have created a way, but it seems it is not perfect.

What could have been useful for me is like an example of a working version of the memory management in the desktop version, using the same feature set as UWP has. I do not understand the internal workings of the JIT and how it handles memory in this project to find what is wrong, and my Paid personal projects came first in time.

Today I don't have that much time for this as back then, but I could help cleaning this up. But actual code for those areas based on the UWP feature set would be really helpful.

**Comment by hrydgard** [comment]
I would love to give some feedback but it's almost impossible to review as-is - Github won't even load that huge commit. I could review through command line git, but it would be much easier if you could split it into smaller steps.

**Comment by perneky** [comment]
A full review at this stage is not necessary. Right now I am not sure that the memory mapping can be done correctly for UWP. That part is actually 2-3 files if I remember correctly. If it turns out that it can not be done correctly for UWP, then I would not put more effort into this.

I would be happy even with a branch of the original project, where on desktop windows the memory management would be done with the functions available for UWP (Virtual memory functions with postfixed with FromApp). Then I would integrate that into this one. If that works and the memory becomes stable, then I would start cleaning this up.

I really would not spend time slicing and cleaning if this memory problem could not be solved.

**Comment by tabnk** [comment]
Interesting. Anyone try it on Xbox one yet ?

**Comment by DarkLegend481** [comment]
Are you using the x32 version? Or the iOS edition?

Because, iOS (I read this) uses ARM architeture... And i tried to use the x32 version (Direct3D9) on my Nokia Lumia. It works, but very slow...

**Comment by tomatelopes** [comment]
Why did you don't try to compile iOS build in UWP appx format with [iOS to UWP Bridge](https://developer.microsoft.com/pt-br/windows/bridges/ios)?

**Comment by Orphis** [comment]
It is very limited and will probably not support all the low level parts an emulator requires.

**Comment by hrydgard** [comment]
While we do want to support UWP, this massive commit is nearly impossible to work with and merge. Until it's split up into smaller commits and rebased, I'm closing this pull request because as it is it's just not feasible to merge it.

**Comment by metalex10** [comment]
x360,Xone and win10uwp is not planing?

**Comment by hrydgard** [comment]
Hi @perneky,

It's time to pick up this thread again - now we have a D3D11 backend so no need for ANGLE trickery or anything like that, so performance will be better.

I'm probably going to start over rather than trying a merge, but I'll probably pick some bits and pieces of this. You don't have to do anything of course, just thought I'd let you know.

**Pull request closed without merging**